src/main/scala/org/hammerlab/guacamole/readsets/rdd/CoverageRDD.scala

```diff
@@ -120,24 +114,55 @@ class CoverageRDD[R <: ReferenceRegion: ClassTag](@transient rdd: RDD[R])
     (depthRuns, map.getOrElse(true, 0), map.getOrElse(false, 0))
   }
 
-  /**
-   * Alternative implementation of this.coverage; will generally perform significantly worse on sorted inputs. Useful as
-   * a sanity check.
-   */
-  def shuffleCoverage(halfWindowSize: Int,
-                      contigLengthsBroadcast: Broadcast[ContigLengths]): RDD[(Position, Coverage)] =
-    (for {
+  private[rdd] def traversalCoverage(halfWindowSize: Int,
+                                     lociBroadcast: Broadcast[LociSet]): RDD[(Position, Coverage)] =
+    rdd
+      .mapPartitions(
+        it =>
+          for {
+
+            // For each contig, and only the regions that lie on it…
+            (contigName, contigRegions) <- ContigsIterator(it.buffered)
+
+            // Iterator over loci overlapped, each with a Coverage recording coverage-depth data.
+            contigCoverages = ContigCoverageIterator(halfWindowSize, contigRegions)
+
+            // Iterator of eligible loci on this contig.
+            lociContig = lociBroadcast.value.onContig(contigName).iterator
+
+            // Filter (locus, coverage) iterator by eligible loci for this contig.
+            (locus, coverage) <- contigCoverages.intersect(lociContig)
+
+          } yield
+            Position(contigName, locus) -> coverage
+      )
+      .reduceByKey(_ + _)
+      .sortByKey()
+
+  private def coveragesForRegion(region: ReferenceRegion,
+                                 halfWindowSize: Int,
+                                 loci: LociSet): Iterator[(Position, Coverage)] = {
+
+    val ReferenceRegion(contigName, start, end) = region
 
-      // For each region…
-      ReferenceRegion(contigName, start, end) <- rdd
+    // Compute the bounds of loci that this region should contribute 1 unit of coverage-depth to.
+    val lowerBound = math.max(0, start - halfWindowSize)
+    val upperBound = end + halfWindowSize
 
-      // Compute the bounds of loci that this region should contribute 1 unit of coverage-depth to.
-      contigLength = contigLengthsBroadcast.value(contigName)
-      lowerBound = math.max(0, start - halfWindowSize)
-      upperBound = math.min(contigLength, end + halfWindowSize)
+    // Iterator over loci spanned by the current region, including the half-window buffer on each end.
+    val readLocusIterator = new LociIterator(Iterator(Interval(lowerBound, upperBound)).buffered)
```

Comment:
nit: mix of `region` and `read` here ... I don't love our use of `region` as our name for things with reference coordinates ... but until there is something better.